cmd/crank/beta/convert/pipelinecomposition/converter.go

```diff
@@ -147,15 +147,98 @@ func processFunctionInput(input *Input) *runtime.RawExtension {
 		"environment": processedInput.Environment.DeepCopy(),
-		"patchSets":   processedInput.PatchSets,
-		"resources":   processedInput.Resources,
+		"patchSets":   MigratePatchPolicy(processedInput.PatchSets),
+		"resources":   MigratePatchPolicyInResources(processedInput.Resources),
 	}
 
 	return &runtime.RawExtension{
 		Object: &unstructured.Unstructured{Object: inputType},
 	}
 }
 
+// MigratePatchPolicyInResources patches resources.
+func MigratePatchPolicyInResources(resources []v1.ComposedTemplate) []ComposedTemplate {
+	composedTemplates := []ComposedTemplate{}
+
+	for _, resource := range resources {
+		composedTemplate := ComposedTemplate{}
+		composedTemplate.ComposedTemplate = resource
+		composedTemplate.Patches = patchPatches(resource.Patches)
+
+		composedTemplate.ComposedTemplate.Patches = nil
+		composedTemplates = append(composedTemplates, composedTemplate)
+	}
+	return composedTemplates
+}
+
+// MigratePatchPolicy migrates mergeoptions.
+func MigratePatchPolicy(patchset []v1.PatchSet) []PatchSet {
+	newPatchSets := []PatchSet{}
+
+	for _, patchSet := range patchset {
+		newpatchset := PatchSet{}
+		newpatchset.Name = patchSet.Name
+		newpatchset.Patch = patchPatches(patchSet.Patches)
+
+		newPatchSets = append(newPatchSets, newpatchset)
+	}
+
+	return newPatchSets
+}
+
+func patchPatches(patches []v1.Patch) []Patch {
+	newPatches := []Patch{}
+
+	for _, patch := range patches {
+		newpatch := Patch{}
+		newpatch.Patch = patch
+
+		if patch.Policy != nil && patch.Policy.MergeOptions != nil {
+			newpatch.Policy = patchPolicy(patch.Policy)
+			newpatch.Patch.Policy = nil
+		}
+
+		newPatches = append(newPatches, newpatch)
+	}
+
+	return newPatches
+}
+
+func patchPolicy(policy *v1.PatchPolicy) *PatchPolicy {
+	if policy.FromFieldPath == nil {
+		return nil
+	}
+
+	pp := &PatchPolicy{
+		FromFieldPath: policy.FromFieldPath,
+	}
+
+	mo := policy.MergeOptions
+	switch {
+	case mo.KeepMapValues == nil && mo.AppendSlice == nil:
+		pp.ToFieldPath = ptr.To(ToFieldPathPolicyForceMergeObjects)
+	case mo.AppendSlice == nil:
+		pp.ToFieldPath = ptr.To(ToFieldPathPolicyMergeObjects)
+	case mo.KeepMapValues == nil:
+		pp.ToFieldPath = ptr.To(ToFieldPathPolicyForceMergeObjectsAppendArrays)
+	default:
+		pp.ToFieldPath = ptr.To(ToFieldPathPolicyMergeObjectsAppendArrays)
+	}
+	return pp
+}
+
+// A ToFieldPathPolicy determines how to patch to a field path.
+type ToFieldPathPolicy string
+
+// ToFieldPathPatchPolicy defines the policy for the ToFieldPath in a Patch.
+const (
+	ToFieldPathPolicyReplace                       ToFieldPathPolicy = "Replace"
+	ToFieldPathPolicyMergeObjects                  ToFieldPathPolicy = "MergeObjects"
+	ToFieldPathPolicyMergeObjectsAppendArrays      ToFieldPathPolicy = "MergeObjectsAppendArrays"
+	ToFieldPathPolicyForceMergeObjects             ToFieldPathPolicy = "ForceMergeObjects"
+	ToFieldPathPolicyForceMergeObjectsAppendArrays ToFieldPathPolicy = "ForceMergeObjectsAppendArrays"
+)
+
```

Comment:
I think it's reasonable to include these in `types.go` also, they are part of the type information.

Reply:
good catch, moved!